File: src/weight.cc
#include "jopenfst.h"

namespace fst::script {

// Weight representation depends on the semiring.
double GetWeight(const WeightClass &weight) {
  if (weight.Type() == "tropical") {
    return weight.GetWeight<fst::TropicalWeight>()->Value();
  } else if (weight.Type() == "log") {
    return weight.GetWeight<fst::LogWeight>()->Value();
  } else if (weight.Type() == "log64") {
    return weight.GetWeight<fst::Log64Weight>()->Value();
  } else {
    FSTERROR() << "weight type not supported in julia";
    return 0.0;
  }
}

// Weight representation depends on the semiring.
WeightClass GetWeightClass(const double weight, 
                           const std::string &weight_type) {
  if (weight_type == "tropical") {
    return WeightClass(fst::TropicalWeight(weight));
  } else if (weight_type == "log") {
    return WeightClass(fst::LogWeight(weight));
  } else if (weight_type == "log64") {
    return WeightClass(fst::Log64Weight(weight));
  } else {
    FSTERROR() << "weight type not supported in julia";
    return WeightClass(fst::TropicalWeight::One());
  }
}

}  // namespace fst::script

File: src/weight.jl
abstract type Weight end
abstract type TropicalWeight <: Weight end
abstract type LogWeight <: Weight end
abstract type Log64Weight <: Weight end

# keyed on arc name
_WeightType = Dict([("standard", TropicalWeight), ("tropical", TropicalWeight),
                    ("log", LogWeight), ("log64", Log64Weight)])

_ArcType = Dict([(TropicalWeight, "standard"), (LogWeight, "log"),
                (Log64Weight, "log64")])

